src/envs/screws.py

```diff
@@ -0,0 +1,509 @@
+        gripper_width = state.get(self.robot, "width")
+        # NOTE: this 1.3 factor is just to make the gripper wide enough so that
+        # even when it grabs a block on its very edge, it doesn't look strange
+        # in the visualization.
+        plt.plot(
+            [(gripper_x - gripper_width / 2.0) + (self._screw_width * 1.3),
+             (gripper_x + gripper_width / 2.0) + (self._screw_width * 1.3)],
+            [gripper_y, gripper_y],
+            color="black")
+        plt.plot(
+            [(gripper_x - gripper_width / 2.0) + (self._screw_width * 1.3),
+             (gripper_x - gripper_width / 2.0) + (self._screw_width * 1.3)],
+            [gripper_y - gripper_width / 4.0, gripper_y],
+            color="black")
+        plt.plot(
+            [(gripper_x + gripper_width / 2.0) + (self._screw_width * 1.3),
+             (gripper_x + gripper_width / 2.0) + (self._screw_width * 1.3)],
+            [gripper_y - gripper_width / 4.0, gripper_y],
+            color="black")
+
+        # Find which object is the goal screw.
+        assert len(task.goal) == 1
+        goal_pred = list(task.goal)[0]
+        goal_screw = goal_pred.objects[0]
+        assert goal_screw.type.name == 'screw'
+
+        # Draw screws
+        for screw in state.get_objects(self._screw_type):
+            screw_x = state.get(screw, "pose_x")
+            screw_y = state.get(screw, "pose_y")
+            screw_width = state.get(screw, "width")
+            screw_height = state.get(screw, "height")
+            if screw == goal_screw:
+                color = 'gold'
+            else:
+                color = 'grey'
+            rect = plt.Rectangle((screw_x, screw_y),
+                                 screw_width,
+                                 screw_height,
+                                 edgecolor=color,
+                                 facecolor=color,
+                                 label=screw.name)
+            ax.add_patch(rect)
+
+        img = utils.fig2data(fig, dpi=150)
+        plt.close()
+        return [img]
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return {
+            self._ScrewPickupable, self._AboveReceptacle, self._HoldingScrew,
+            self._ScrewInReceptacle
+        }
+
+    @property
+    def goal_predicates(self) -> Set[Predicate]:
+        return {self._ScrewInReceptacle}
+
+    @property
+    def types(self) -> Set[Type]:
+        return {self._screw_type, self._gripper_type, self._receptacle_type}
+
+    @property
+    def options(self) -> Set[ParameterizedOption]:
+        return {
+            self._MoveToScrew, self._MoveToReceptacle, self._MagnetizeGripper,
+            self._DemagnetizeGripper
+        }
+
+    @property
+    def action_space(self) -> Box:
+        # dimensions: [dx, dy, magnetized vs. unmagnetized].
+        lowers = np.array([
+            -(self.rz_x_ub - self.rz_x_lb), -(self.rz_y_ub - self.rz_y_lb), 0.0
```

Comment:
suggest making variables like `rz_width = self.rz_x_ub - self.rz_x_lb` and reusing in `lowers` and `uppers`